src/shared/bootspec.c

```diff
@@ -1046,6 +1052,189 @@ int boot_config_augment_from_loader(
         return 0;
 }
 
+static void split_last_three_by_slash(char *src, char **preprelast, char **prelast, char **last) {
+        char *p, *saveptr;
+
+        *last = *prelast = *preprelast = NULL;
+        if (!src)
+            return;
+        for(; ; src = NULL) {
+                p = strtok_r(src, "/", &saveptr);
+                if (!p)
+                    break;
+
+                *preprelast = *prelast;
+                *prelast = *last;
+                *last = p;
+
+        }
+}
+
+/*
+ * check if two dev_t are partitions of the same disk
+ */
+static int check_parent_by_devid(dev_t a, dev_t b) {
+        char sysfsname[2048], bufa[2048], bufb[2048];
+        int r;
+        char *last, *prelast, *preprelast, *aprelast;
+
+        snprintf(sysfsname, sizeof(sysfsname) - 1, "/sys/dev/block/%d:%d", major(a), minor(a));
+        r = readlink(sysfsname, bufa, sizeof(bufa) -1);
+        if (r < 0)
+                return r;
+
+        /* bufa is in the form of '../../devices/pci0000:00/0000:00:01.2/0000:01:00.1/ata4/host3/target3:0:0/3:0:0:0/block/sde/sde3' */
+        split_last_three_by_slash(bufa, &preprelast, &prelast, &last);
+        if (strcmp(preprelast, "block"))
+                return -ENOENT;
+
+        aprelast = prelast;
+
+        snprintf(sysfsname, sizeof(sysfsname) - 1, "/sys/dev/block/%d:%d", major(b), minor(b));
+        r = readlink(sysfsname, bufb, sizeof(bufb) -1);
+        if (r < 0)
+                return r;
+
+        /* bufb is in the form of '../../devices/pci0000:00/0000:00:01.2/0000:01:00.1/ata4/host3/target3:0:0/3:0:0:0/block/sde/sde3' */
+        split_last_three_by_slash(bufb, &preprelast, &prelast, &last);
+        if (strcmp(preprelast, "block"))
+                return -ENOENT;
+
+        return strcmp(prelast, aprelast);
+}
```

Comment:
Use our block_get_whole_disk(), we already have that.

Reply:
Many thanks for the suggestion